csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Assets.cs

```diff
@@ -0,0 +1,232 @@
+
+            if (references is null)
+            {
+                progressMonitor.LogDebug("No references found in the targets section in the assets file.");
+                return Array.Empty<string>();
+            }
+
+            // Find all the compile dependencies for each reference and
+            // create the relative path to the dependency.
+            var dependencies = references
+                .SelectMany(r =>
+                {
+                    var info = r.Value;
+                    var name = r.Key.ToLowerInvariant();
+                    if (info.Type != "package")
+                    {
+                        return Array.Empty<string>();
+                    }
+
+                    // If this is a .NET framework reference then include everything.
+                    return netFrameworks.Any(framework => name.StartsWith(framework))
+                        ? new[] { name }
+                        : info
+                            .Compile
+                            .Select(p => Path.Combine(name, ParseFilePath(p.Key)));
+                })
+                .ToList();
+
+            return dependencies;
+        }
+
+        /// <summary>
+        /// Gets the framework dependencies from the assets file.
+        ///
+        /// Example:
+        /// "project": {
+        //     "version": "1.0.0",
+        //     "frameworks": {
+        //         "net7.0": {
+        //             "frameworkReferences": {
+        //                 "Microsoft.AspNetCore.App": {
+        //                     "privateAssets": "none"
+        //                 },
+        //                 "Microsoft.NETCore.App": {
+        //                     "privateAssets": "all"
+        //                 }
+        //             }
+        //         }
+        //     }
+        // }
+        //
+        /// Returns { "microsoft.aspnetcore.app.ref", "microsoft.netcore.app.ref" }
+        /// </summary>
+        private IEnumerable<string> GetFrameworkDependencies(JObject json)
+        {
+
+            var frameworks = json
+                .GetProperty("project")?
+                .GetProperty("frameworks");
+
+            if (frameworks is null)
+            {
+                progressMonitor.LogDebug("No framework section in assets.json.");
+                return Array.Empty<string>();
+            }
+
+            // If there are more than one framework we need to pick just one.
+            // For now, we just pick the first one.
+            var references = frameworks
+                .GetFirstPropertyValue()?
+                .GetProperty("frameworkReferences");
+
+            if (references is null)
+            {
+                progressMonitor.LogDebug("No framework references in assets.json.");
+                return Array.Empty<string>();
+            }
+
+            var dependencies = references
+                .Properties()
+                .Select(f => $"{f.Name}.Ref".ToLowerInvariant())
```

Comment:
Is this `.Ref` concatenation always okay?

Reply:
I have not seen evidence of otherwise, but I can't say for sure either.
If we point to a folder that doesn't exist it will thrown an exception when the assembly cache is created (basically I wanted to see if would run into this issue).
We could also make the assembly cache a bit less strict: That is, just make it a best effort attempt to load all the paths that we provide.

Reply:
I think we shouldn't fail completely the dependency fetching. We could log a warning, and still continue.